File: vertx-testframework/src/main/java/org/vertx/java/framework/TestVerticle.java
package org.vertx.java.framework;

import org.vertx.java.deploy.Verticle;

public abstract class TestVerticle extends Verticle {
    public TestUtils testUtils;

    public final void start() {
        testUtils = new TestUtils(vertx);
        doStart();

        getContainer().getLogger().info(this + " started");
        testUtils.registerTests(this);

        testUtils.appReady();
    }

    public void stop() {
        doStop ();

        getContainer().getLogger().info(this + " stopped");
        testUtils.unregisterAll();
        testUtils.appStopped();
    }

    protected void doStart () {}

    protected void doStop () {}
}

File: vertx-testframework/src/main/kotlin/org/vertx/kotlin/framework/TestVerticle.kt
package org.vertx.kotlin.framework

import org.vertx.java.deploy.Verticle
import org.vertx.java.framework.TestUtils

import org.vertx.kotlin.core.*

public abstract class TestVerticle() : Verticle() {
    private var _testUtils: TestUtils? = null

    // FIXME: kotlin compiler bug (no bridge when called from closure)
    public val testUtils: TestUtils
        get() {
            if (_testUtils == null) {
                _testUtils = TestUtils(vertx)
            }
            return _testUtils!!
        }

    public override final fun start() {
        doStart()

        logger.info("$this started")
        testUtils.registerTests(this)

        testUtils.appReady()
    }

    public override fun stop() {
        doStop ()

        logger.info("$this stopped")
        testUtils.unregisterAll()
        testUtils.appStopped()
    }

    protected open fun doStart () {}

    protected open fun doStop () {}
}